sklearn/preprocessing/_encoders.py

```diff
@@ -795,10 +809,17 @@ def transform(self, X):
         -------
         X_out : sparse matrix or a 2-d array
             Transformed input.
-
+            
         """
-        X_int, _ = self._transform(X)
-        return X_int.astype(self.dtype, copy=False)
+
+        X_int, _ = self._transform(X, encode_missing=self._encode_missing)
+
+        transformed = X_int.astype(self.dtype, copy=False)
+
+        if not self._encode_missing:
+            transformed[transformed == -1] = np.nan
```

Comment:
This will not work for eg `dtype='int64'`. 

But not sure what best to do about this. We raise an informative error if `dtype` is not floating, and the encoding is set to passthrough NaNs